### PR TITLE
COL-460 Preview image support for video files

### PR DESCRIPTION
**Pull request by paulkerschen**
https://jira.ets.berkeley.edu/jira/browse/COL-460

When faced with a video file: 
- extract the first frame and process it as it would any other image;
- if faced with a weird encoding, convert to something more standard.

### CODE REVIEW

**Comment by paulkerschen** [comment]
See also https://github.com/ets-berkeley-edu/suitec/pull/291

File: .ebextensions/04_install_ffmpeg.config [hunk 1]
```diff
@@ -0,0 +1,11 @@
+commands:
+  01_become_root:
+    command: 'sudo su'
+  02_get_ffmpeg_executable:
+    command: 'wget -N https://johnvansickle.com/ffmpeg/releases/ffmpeg-release-64bit-static.tar.xz -O ffmpeg-release-64bit-static.tar.xz'
```
Comment (paulkerschen) [inline .ebextensions/04_install_ffmpeg.config]: A public service providing more recent static builds of ffmpeg than the ffmpeg website itself does. Compiling this package from source takes unreasonably long.

File: lib/preview/constants.js [hunk 2]
```diff
@@ -102,6 +102,14 @@ module.exports = {
...
     'PDF': [
       'application/pdf'
+    ],
+    'VIDEO': [
+      'video/mp4',
+      'video/mpeg',
+      'video/quicktime',
+      'video/x-m4v',
+      'video/x-ms-wmv',
+      'video/x-msvideo'
```
Comment (paulkerschen) [inline lib/preview/constants.js]: A sampling of formats used to date in SuiteC, which we're reasonably confident that ffmpeg can handle out of the box.

Comment (johncrossman) [inline lib/preview/constants.js]: @paulkerschen looks right to me. No need for `video/avi`? `video/x-flv`?

Comment (paulkerschen) [inline lib/preview/constants.js]: @johncrossman Apparently not... see https://jira.ets.berkeley.edu/jira/browse/COL-460 for what's been uploaded in the wild to date.

File: lib/preview/image.js [hunk 3]
```diff
@@ -78,6 +81,7 @@ var processImage = module.exports.processImage = function(ctx, file, callback) {
  * @param  {Object}     callback.err          An error object, if any
  * @param  {Result}     callback.imagePath    The path to the large image on disk
  * @param  {Number}     callback.width        The width of the generated image
+ * @param  {Number}     callback.height      The height of the generated image
```
Comment (johncrossman) [inline lib/preview/image.js]: @paulkerschen Space needed above.

Comment (paulkerschen) [inline lib/preview/image.js]: Thanks @johncrossman, amended.

File: lib/preview/video.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+ * @param  {Function}   callback                     Standard callback function
+ * @param  {Object}     callback.err                 An error object, if any
+ * @param  {String}     callback.imagePath           The path to the extracted image
+ */
+var extractFirstFrame = function(ctx, callback) {
+  var extractCommand = [
+    'ffmpeg',
+    util.format('-i "%s"', ctx.sourcePath),
+    '-vframes 1',
+    path.join(ctx.directory, 'frame%d.png')
```
Comment (johncrossman) [inline lib/preview/video.js]: @paulkerschen what determines `%d`?

Comment (paulkerschen) [inline lib/preview/video.js]: @johncrossman Frame number; for some reason ffmpeg requires this format even if it's only outputting one frame. `frame1.png` (see below) is the result.

File: lib/preview/video.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  ].join(' ');
+
+  exec(extractCommand, {'timeout': 200000}, function(err, stdout, stderr) {
+    if (err) {
+      return callback({'code': 500, 'msg': 'Unable to generate image from video frame'});
```
Comment (johncrossman) [inline lib/preview/video.js]: @paulkerschen is there value in appending `err.message` to msg above (and below)?

**Comment by johncrossman** [comment]
I'm done. Very well done, @paulkerschen  One question awaits your answer then I'll merge.

File: lib/preview/video.js (hunk 5 above)
Comment (paulkerschen) [inline lib/preview/video.js]: @johncrossman Yes indeed; amended to include those messages.

File: lib/preview/video.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+  exec(extractCommand, {'timeout': 200000}, function(err, stdout, stderr) {
+    if (err) {
+      var msg = util.format('Error while extracting first frame image from video: %s', err.message);
+      return callback({'code': 500, 'msg': msg});
+    }
+
+    var imagePath = path.join(ctx.directory, 'frame1.png');
+    if (!fs.existsSync(imagePath)) {
+      return callback({'code': 500, 'msg': 'An image file could not be generated'});
```
Comment (johncrossman) [inline lib/preview/video.js]: @paulkerschen One more q: would `imagePath` be useful in msg above? and `convertedVideoPath` at line 143, below? Sorry for nit-pickiness. :)

Comment (paulkerschen) [inline lib/preview/video.js]: @johncrossman A good log line in time saves nine. Amended with paths added.

**Pull request merged**